src/ConnectedMode/Migration/VsAwareFileSystem.cs

```diff
@@ -58,52 +61,148 @@ internal class VsAwareFileSystem : IVsAwareFileSystem
             var content = fileSystem.File.ReadAllText(filePath);
             return Task.FromResult(content);
         }
 
-        public Task SaveAsync(string filePath, string text)
+        public async Task SaveAsync(string filePath, string text)
         {
-            // TODO - error handling
+            Debug.Assert(batchStarted, "Expecting the changes to happen in a batch");
+
             logger.LogMigrationVerbose($"Saving file: {filePath}");
+
+            bool success = await CheckoutForEditAsync(filePath);
+
+            if (!success)
+            {
+                var message = string.Format(MigrationStrings.VSFileSystem_Error_FailedToCheckOutFile, filePath);
+                throw new InvalidOperationException(message);
+            }
+
             fileSystem.File.WriteAllText(filePath, text);
-            return Task.CompletedTask;
         }
 
         public async Task BeginChangeBatchAsync()
         {
+            Debug.Assert(!batchStarted, "Not expecting to already be in a batch");
+
             logger.LogMigrationVerbose("Beginning batch of file changes...");
 
-            queryFileOperation = await GetSccServiceAsync();
+            sccService = await GetSccServiceAsync();
+            sccService.BeginQuerySaveBatch();
+            batchStarted = true;
         }
 
         public Task EndChangeBatchAsync()
         {
-            logger.LogMigrationVerbose("File changes complete");
+            logger.LogMigrationVerbose("Batch file changes completed");
+            sccService.EndQuerySaveBatch();
+            batchStarted = false;
             return Task.CompletedTask;
         }
 
         private async Task<IVsQueryEditQuerySave2> GetSccServiceAsync()
         {
             IVsQueryEditQuerySave2 sccService = null;
-            await threadHandling.RunOnUIThread( () =>
+            await threadHandling.RunOnUIThread(() =>
             {
                 sccService = serviceProvider.GetService(typeof(SVsQueryEditQuerySave)) as IVsQueryEditQuerySave2;
             });
 
             return sccService;
         }
+
+        private async Task CheckOutFilesToDeleteAsync(string folderPath)
+        {
+            // Can't check out the directory for saving - if you do, VS will pop up a dialogue,
+            // even if the "silent" flag is set.
+            bool success = await CheckoutForEditAsync(folderPath);
+
+            if (success)
+            {
+                var files = fileSystem.Directory.GetFiles(folderPath, "*.*", System.IO.SearchOption.AllDirectories);
+                success = await CheckoutForSaveAsync(files);
+            }
+
+            if (!success)
+            {
+                var message = string.Format(MigrationStrings.VsFileSystem_Error_FailedToCheckOutFolderForDeletion, folderPath);
+                throw new InvalidOperationException(message);
+            }
+        }
+
+        private async Task<bool> CheckoutForEditAsync(params string[] fileNames)
```

Comment:
This was copied from the [equivalent method](https://github.com/SonarSource/sonarlint-visualstudio/blob/03123f51a5f22aa2bd6e347f00a33ca01e9ee44b/src/Integration/LocalServices/SourceControlledFileSystem.cs#L133) of the existing `SourceControlledFileSystem` class.

I've tweaked it slightly to make it `async` so we can force the SCC operation to happen on the UI thread (VS crashes if it is called on a background thread),

The legacy `SourceControlledFileSystem` implementation assumed that the caller would ensure that it was called on the main thread.